### PR TITLE
Update guessing sheet

### PR DESCRIPTION
**Pull request by peternga**
* Supports updating guessing sheet with clue word based on what the clue giver would see
* Supports progressing the game by denoting if every player is "ready"
* Allows players to advance their visible letter

### CODE REVIEW

File: src/client/scripts/scenes/gameScene.ts [hunk 1]
```diff
@@ -175,6 +174,28 @@ export default class GameScene extends Phaser.Scene {
       );
     });
 
+    this.socket.on(E.LetterOrdering, (letterordering) => {
```
Comment (nathanleiby) [inline src/client/scripts/scenes/gameScene.ts:177]: letter ordering could have an explicit type

```
(letterOrdering: EType[E.LetterOrdering])
```

File: src/server/lib/gameState.ts [hunk 2]
```diff
@@ -59,11 +66,28 @@ export class ServerGameState {
...
+    return "";
+  }
+
   resetVotesAndClues() {
     this.clues = {};
     this.votes = {};
   }
 
+  areAllPlayersReady(): boolean {
+    return this.playersReady.size >= Array.from(this.players.keys()).length;
```
Comment (nathanleiby) [inline src/server/lib/gameState.ts:84]: nit: `Array.from(this.players.keys())` is also available as `getPlayerIDs()`

File: src/server/lib/scenes/game.ts [hunk 3]
```diff
@@ -1,9 +1,42 @@
 import * as _ from "lodash";
 
 import { ServerGameState } from "../../lib/gameState";
+import { PlayStateEnum, PlayStates } from "../../../shared/constants";
 import { E, EType } from "../../../shared/events";
 import { playerID } from "../playerUtils";
 
+const getLetterOrdering = (gameState) => {
```
Comment (nathanleiby) [inline src/server/lib/scenes/game.ts:8]: could add add type to `gameState` => `gameState: ServerGameState`

File: src/shared/constants.ts [hunk 4]
```diff
@@ -79,4 +79,30 @@ export const BaseNPCCards = 7;
...
+  INTERPRET_HINT = "interpret_hint",
+  CHECK_END_CONDITION = "check_end_condition",
+}
+
+export const PlayStates = [
+  PlayStateEnum.DISCUSS,
+  PlayStateEnum.PROVIDE_HINT,
+  PlayStateEnum.INTERPRET_HINT,
+  PlayStateEnum.CHECK_END_CONDITION,
+];
```
Comment (nathanleiby) [inline src/shared/constants.ts:108]: 👍 thanks for adding these enums

File: src/server/lib/gameState.ts [hunk 5]
```diff
@@ -36,8 +37,11 @@ export class ServerGameState {
   deck: Letter[];
   // redTokens,
   // greenTokens
-  clues: { [playerID: string]: Clue };
+  clues: { [playerID: string]: FullClue };
```
Comment (nathanleiby) [inline src/server/lib/gameState.ts:40]: Do we need both Clue and FullClue?

File: src/client/scripts/scenes/gameScene.ts (hunk 1 above)
Comment (peternga) [reply src/client/scripts/scenes/gameScene.ts:177]: Ah yeah. It was easier for me to figure out what I needed first and then apply types after but I think there were a couple places I missed.

File: src/server/lib/gameState.ts (hunk 5 above)
Comment (peternga) [reply src/server/lib/gameState.ts:40]: I think we do. The client shouldn't have access to the word in all cases.

**Pull request merged**